[Digital Identity]: navigator.identity should be implemented by a separate IdentityCredentialManager

marcoscaceres opened the pull request:
#### 2dbfef33bb52c31bed14b8fadf209338031552a5
<pre>
[Digital Identity]: navigator.identity should be implemented by a separate IdentityCredentialManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=269165">https://bugs.webkit.org/show_bug.cgi?id=269165</a>
<a href="https://rdar.apple.com/123184696">rdar://123184696</a>

Reviewed by Andy Estes.

Implements IdentityCredentialsContainer on top of CredentialsContainer by
virtualizing some of the methods of CredentialsContainer.

For IdentityCredentialsContainer, we implement the logic for:

 * navigator.identity.get()
 * navigator.identity.create() - Default to just return null.

And we defer the logic for the following to CredentialsContainer:
 * navigator.identity.store() - which just rejects with a NotSupportedError.

Adds performCommonChecks() to CredentialsContainer to DRY up the code,
and so that we can use it in IdentityCredentialsContainer.

And the DigitalCredentialsEnable preference is now set to &quot;testable&quot;.

* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https.html: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-create-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-create-basics.https.html: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-store-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-store-basics.https.html: Added.
* LayoutTests/http/wpt/identity/idl.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/idl.https.html: Added.
* LayoutTests/http/wpt/identity/setDigitalCredentialsEnable.https.html:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/credentialmanagement/CredentialRequestOptions.h:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::get):
(WebCore::CredentialsContainer::isCreate):
(WebCore::CredentialsContainer::performCommonChecks):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
(WebCore::CredentialsContainer::create):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl:
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp: Added.
(WebCore::IdentityCredentialsContainer::IdentityCredentialsContainer):
(WebCore::IdentityCredentialsContainer::get):
(WebCore::IdentityCredentialsContainer::isCreate):
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.h: Copied from Source/WebCore/Modules/identity/NavigatorIdentity.h.
* Source/WebCore/Modules/identity/NavigatorIdentity.cpp:
(WebCore::NavigatorIdentity::identity):
* Source/WebCore/Modules/identity/NavigatorIdentity.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::defaults):

Canonical link: <a href="https://commits.webkit.org/275939@main">https://commits.webkit.org/275939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/617b23372d15129e7c4112b66a36c1161b0bd207

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39340 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45520 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19662 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35736 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19294 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37256 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16729 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16882 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38308 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1269 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/36673 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39421 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47392 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42840 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18129 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42528 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19666 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41186 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9640 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19844 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49850 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19298 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10069 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -0,0 +1,80 @@
+        return;
+    }
+
+    // TODO: Implement the actual get logic.
```
FIXME*

annevk commented on Source/WebCore/Modules/identity/NavigatorIdentity.cpp:
```diff
@@ -46,9 +46,9 @@ ASCIILiteral NavigatorIdentity::supplementName()
+        m_credentialsContainer = IdentityCredentialsContainer::create(WTFMove(document));
 
-    return m_credentialsContainer.get();
+    return static_cast<CredentialsContainer*>(m_credentialsContainer.get());
```
downcast? Or maybe even checkedDowncast.

marcoscaceres replied:
I wasn't sure... from what I read "on the internets"™️, `static_cast` seemed an appropriate here because `IdentityCredentialsContainer` is guaranteed. @aprotyas, do you know which is correct? They are both used all over the WebKit codebase.

aestes replied:
Right, this is an upcast, not a downcast. Since `IdentityCredentialsContainer` is a `CredentialsContainer`, the `static_cast` shouldn't even be necessary.

aprotyas replied:
Yes as @aestes said, pretty sure you don't need any cast here?

annevk reviewed (commented):
Looks okay overall to me, but don't feel like I know this well enough to r+.

aestes commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
```diff
@@ -76,16 +76,9 @@ void CredentialsContainer::get(CredentialRequestOptions&& options, CredentialPro
     // with enhancement from 14 November 2017 Editor's Draft.
-    if (!m_document || !m_document->page()) {
-        promise.reject(Exception { ExceptionCode::NotSupportedError });
+    if (!performCommonChecks(std::forward<CredentialRequestOptions>(options), promise)) {
```
WTFMove instead of std::forward

aestes commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
```diff
@@ -113,16 +106,8 @@ void CredentialsContainer::isCreate(CredentialCreationOptions&& options, Credent
-    }
-    if (options.signal && options.signal->aborted()) {
-        promise.reject(Exception { ExceptionCode::AbortError, "Aborted by AbortSignal."_s });
+    if (!performCommonChecks(std::forward<CredentialCreationOptions>(options), promise))
```
Ditto

aestes commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
```diff
@@ -144,6 +129,22 @@ void CredentialsContainer::preventSilentAccess(DOMPromiseDeferred<void>&& promis
 }
 
+template<typename Options>
+bool CredentialsContainer::performCommonChecks(Options&& options, CredentialPromise& promise)
```
It would be simpler to just pass a `const Options&` here. Then there's no need to move/forward at the call sites.

marcoscaceres replied:
Agree... and yeah, was silly I was poking at things after moving them.

aestes commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
```diff
@@ -39,31 +39,35 @@ enum class Scope;
-private:
     CredentialsContainer(WeakPtr<Document, WeakPtrImplWithEventTargetData>&&);
 
+    virtual ~CredentialsContainer() { }
```
```suggestion
    virtual ~CredentialsContainer() = default;
```

aestes commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -0,0 +1,80 @@
+    if (!performCommonChecks(std::forward<CredentialRequestOptions>(options), promise))
+        return;
+
+    if (!options.digital) {
```
Since you forwarded `options` above it's no longer valid to access it here. That's why `performCommonChecks` should just take a `const Options&`.

marcoscaceres replied:
🤦 of course.

aestes commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.h:
```diff
@@ -0,0 +1,58 @@
+
+class IdentityCredentialsContainer final : public CredentialsContainer {
+public:
+    static Ref<IdentityCredentialsContainer> create(WeakPtr<Document, WeakPtrImplWithEventTargetData>&& document) { return adoptRef(*new IdentityCredentialsContainer(WTFMove(document))); }
```
Nit: write this like a normal function with the braces on their own lines.

aestes commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.h:
```diff
@@ -0,0 +1,58 @@
+    void isCreate(CredentialCreationOptions&&, CredentialPromise&&) override;
+
+private:
+    IdentityCredentialsContainer(WeakPtr<Document, WeakPtrImplWithEventTargetData>&&);
```
```suggestion
    explicit IdentityCredentialsContainer(WeakPtr<Document, WeakPtrImplWithEventTargetData>&&);
```

marcoscaceres commented on LayoutTests/http/wpt/identity/idl.https-expected.txt:
```diff
@@ -0,0 +1,88 @@
+PASS CredentialsContainer interface: navigator.identity must inherit property "create(optional CredentialCreationOptions)" with the proper type
+PASS CredentialsContainer interface: calling create(optional CredentialCreationOptions) on navigator.identity with too few arguments must throw TypeError
+PASS CredentialsContainer interface: navigator.identity must inherit property "preventSilentAccess()" with the proper type
+FAIL PasswordCredential interface: existence and properties of interface object assert_own_property: self does not have own property "PasswordCredential" expected property "PasswordCredential" missing
```
These FAILs are expected, as we don't implement PasswordCredential or FederatedCredential.

marcoscaceres commented on LayoutTests/http/wpt/identity/idl.https-expected.txt:
```diff
@@ -0,0 +1,88 @@
+PASS DigitalCredential interface: existence and properties of interface prototype object
+PASS DigitalCredential interface: existence and properties of interface prototype object's "constructor" property
+PASS DigitalCredential interface: existence and properties of interface prototype object's @@unscopables property
+FAIL DigitalCredential interface: attribute protocol assert_true: The prototype object must have a property "protocol" expected true got false
```
I'll fix this as a followup.

marcoscaceres replied:
https://github.com/WebKit/WebKit/pull/25697

marcoscaceres commented on LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
```diff
@@ -11,6 +11,7 @@ navigator.contacts is OK
 navigator.cookieEnabled is OK
 navigator.credentials is OK
 navigator.hardwareConcurrency is OK
+navigator.identity is OK
```
`identity` is also showing up in mac-wk1, unfortunately. I need to figure out how to turn that off. Any suggestions?

aestes commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void IdentityCredentialsContainer::get(CredentialRequestOptions&& options, CredentialPromise&& promise)
+{
+    if (!performCommonChecks(std::forward<CredentialRequestOptions>(options), promise))
```
No need to `std::forward` here anymore.

marcoscaceres replied:
Sent https://github.com/WebKit/WebKit/pull/25752

aestes commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void IdentityCredentialsContainer::isCreate(CredentialCreationOptions&& options, CredentialPromise&& promise)
+{
+    if (!performCommonChecks(std::forward<CredentialCreationOptions>(options), promise))
```
Ditto

marcoscaceres replied:
Oh whoops. I thought I got rid of these. 

Sent https://github.com/WebKit/WebKit/pull/25752

marcoscaceres committed:
[Digital Identity]: navigator.identity should be implemented by a separate IdentityCredentialManager
https://bugs.webkit.org/show_bug.cgi?id=269165
rdar://123184696

Reviewed by Andy Estes.

Implements IdentityCredentialsContainer on top of CredentialsContainer by
virtualizing some of the methods of CredentialsContainer.

For IdentityCredentialsContainer, we implement the logic for:

 * navigator.identity.get()
 * navigator.identity.create() - Default to just return null.

And we defer the logic for the following to CredentialsContainer:
 * navigator.identity.store() - which just rejects with a NotSupportedError.

Adds performCommonChecks() to CredentialsContainer to DRY up the code,
and so that we can use it in IdentityCredentialsContainer.

And the DigitalCredentialsEnable preference is now set to "testable".

* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https.html: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-create-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-create-basics.https.html: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-store-basics.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/identtycredentialscontainer-store-basics.https.html: Added.
* LayoutTests/http/wpt/identity/idl.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/idl.https.html: Added.
* LayoutTests/http/wpt/identity/setDigitalCredentialsEnable.https.html:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/credentialmanagement/CredentialRequestOptions.h:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::get):
(WebCore::CredentialsContainer::isCreate):
(WebCore::CredentialsContainer::performCommonChecks):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
(WebCore::CredentialsContainer::create):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.idl:
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp: Added.
(WebCore::IdentityCredentialsContainer::IdentityCredentialsContainer):
(WebCore::IdentityCredentialsContainer::get):
(WebCore::IdentityCredentialsContainer::isCreate):
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.h: Copied from Source/WebCore/Modules/identity/NavigatorIdentity.h.
* Source/WebCore/Modules/identity/NavigatorIdentity.cpp:
(WebCore::NavigatorIdentity::identity):
* Source/WebCore/Modules/identity/NavigatorIdentity.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::defaults):

Canonical link: https://commits.webkit.org/275939@main
